### PR TITLE
Random.self_init with getentropy

### PR DESCRIPTION
**Pull request by mndrix**
`Random.self_init` currently uses `/dev/urandom` to acquire entropy.  `/dev/urandom` may not be available in chroot environments or in some containers. It's also subject to file handle exhaustion. To work around these problems, many systems have introduced system calls which supply high-quality random numbers. They're designed to seed a process-local PRNG, as we do in `Random.self_init`.

This draft pull request teaches `caml_unix_random_seed` to try `getentropy` before falling back to `/dev/urandom`. There's some incompatibility across platforms, but it's not bad:

- [Linux](https://manpages.debian.org/bullseye/manpages-dev/getentropy.3.en.html), [FreeBSD](https://www.freebsd.org/cgi/man.cgi?query=getentropy&apropos=0&sektion=0&manpath=FreeBSD+13.0-RELEASE+and+Ports&arch=default&format=html), [OpenBSD](https://man.openbsd.org/OpenBSD-7.0/getentropy), [Solaris](https://docs.oracle.com/cd/E88353_01/html/E37841/getentropy-2.html), and [Illumos](https://illumos.org/man/3c/getentropy) declare `getentropy` in `unistd.h`
- [macOS](https://keith.github.io/xcode-man-pages/getentropy.2.html) declares `getentropy` in `sys/random.h`
- [NetBSD](https://man.netbsd.org/arc4random.3) doesn't have `getentropy` (it has `arc4random_buf` but this PR doesn't use it)

This PR works on the machines that I have access to: Ubuntu 20.04 LTS, macOS 10.15.7, and OpenBSD 7.0 (with #10875 applied).

Outstanding questions:
- is it worth `#include <unistd.h>` when `HAS_GETENTROPY`? If a system `HAS_GETENTROPY`, it almost certainly `HAS_UNISTD` so it will have already done the import
- should I add support for NetBSD via `arc4random_buf`?
- if `getentropy` fails, is it worth falling back to `/dev/urandom`.  `getentropy` is designed not to fail unless things are really bad, in which case `/dev/urandom` seems unlikely to work either. I guess it doesn't hurt anything to try both

### CODE REVIEW

**Comment by gasche** [comment]
This seems reasonable to me, but the header-selection looks a bit fragile (isn't there a risk that `getentropy` is found, but not in the headers we are thinking of, and so the build fails?). Would it be reasonably easy to check that we actually know which header to include (to use `getentropy`) at configure-time? (I know very little about configure/autoconf, but I think it can perform some small compilation tests.)

File: runtime/sys.c
Comment (xavierleroy) [inline runtime/sys.c]: Watch out: `data` is an array of integers... If you look at the /dev/urandom case below, the 12 random bytes are expanded to 12 integers.  It was necessary with the previous implementation of Random, based on a lagged Fibonacci generator, and is no longer strictly necessary, but it would be nicer to stay consistent with the /dev/urandom case.

File: configure.ac
Comment (xavierleroy) [inline configure.ac]: I'd rather use `AC_CHECK_DECL` to check for a visible declaration in `<unistd.h>` or in the Apple-specific header.  Under Linux for instance, `getentropy` is not declared if the user compiles in strict ISO C mode or in POSIX_C_SOURCE mode.

**Review by xavierleroy — COMMENTED** [review]
Nice suggestion, thank you.

I was thinking of revising "self-initialization" to start with a string rather than an array of integers.  Adding `getentropy` support would fit this revision well.

**Comment by avsm** [comment]
@mndrix wrote:
> should I add support for NetBSD via arc4random_buf?

There's chatter in [NetBSD tech-userlevel](https://mail-index.netbsd.org/tech-userlevel/2020/05/02/msg012333.html) about adding *getentropy(3)* since it has been [added to the latest POSIX standard](http://mail-index.netbsd.org/tech-security/2020/10/16/msg001099.html).  So probably worth holding off on arc4random_buf as the existing /dev/urandom works on NetBSD in the meanwhile.

File: runtime/sys.c
Comment (damiendoligez) [reply runtime/sys.c]: The comment on lines 587 is wrong: we have 96 **bytes** (768 bits) of randomness, not 96 **bits**. Assuming 64-bit `intnat`.

Indeed this looks overkill (the PRNG has 256 bits of internal state). Still, the /dev/urandom code below provides only 96 bits, and it would be better to provide 256.

Maybe the /dev/urandom code should also be fixed to provide 256 bits.

Comment (xavierleroy) [reply runtime/sys.c]: No.  If you read 12 good random bytes, then expand each byte into a 64-bit integer, you still only have 12 * 8 = 96 bits of entropy.

Comment (xavierleroy) [reply runtime/sys.c]: `getentropy` takes a `void *` and a count in bytes.  Here we're passing an array of 64-bit integers but it still fills only the first integer and half of the second integer.  This isn't great, and that's why it should better be expanded like the /dev/urandom read does.

Comment (mndrix) [reply runtime/sys.c]: Thank you. I didn't notice the type differences there.

**Comment by mndrix** [comment]
I've pushed a new version of this PR, and I'm much happier with it.

> the header-selection looks a bit fragile

@gasche Good point.  I followed @xavierleroy suggestion and used `AC_CHECK_DECL` to verify that
`unistd.h` declares `getentropy`.

By incorporating @avsm observation about `getentropy` in upcoming POSIX, I was able to remove the need for  special header handling: `HAS_UNISTD` already covers it. The cost is that we lose support for `getentropy` for Apple.  I felt that the reduction in complexity was worth that cost; same for NetBSD, in line with @avsm.

> it would be nicer to stay consistent with the /dev/urandom case.

@xavierleroy the new PR treats `getentropy` and `/dev/urandom` much more consistently.

I think this is ready for further review at anyone's convenience.

**Review by xavierleroy — APPROVED** [review]
This looks very good to me.  Thanks!

I took the liberty to push an additional commit to this PR, re-adding support for macOS and falling back to /dev/urandom if getentropy() fails.

For macOS, I think it's worth a special case: the man page for getentropy() mentions that it is supported even in jails and sandboxes, while /dev/urandom generally is not, so that's a good reason to use getentropy().

For /dev/urandom fallback: old Linux kernels (as found in RHEL/CentOS 7 for example) don't support the undelying getrandom() syscall,  while a more recent glibc will declare getentropy(), so falling back is better in this admittedly very unlikely case.  99.99% of the time, getentropy() will succeed and the fallback won't take place.

Let me know what you think.

**Comment by xavierleroy** [comment]
Before merging this, I'd like to run a round of "Inria CI precheck", meaning that Inria CI needs to be fixed to work for 5.00 first.

On an unrelated note, it looks like Cygwin implements getentropy too.

**Comment by mndrix** [comment]
> For macOS, I think it's worth a special case

That works for me.

> old Linux kernels ... while a more recent glibc

Good catch. I couldn't think of a scenario where that might happen, but you found one.

> Let me know what you think.

Overall, I like it. The localized `fd` is better than what I had. The `else` hiding inside an `#ifdef` puzzled me for a minute. I eventually convinced myself that it was correct. I'm not familiar enough with C to know if that's a common idiom that others will recognize easily.

**Comment by xavierleroy** [comment]
> The else hiding inside an #ifdef puzzled me for a minute. I eventually convinced myself that it was correct. I'm not familiar enough with C to know if that's a common idiom that others will recognize easily.

It's a horrible hack, and I actually got it wrong (the fallthrough code was not there, the `#if / #else / #endif` should have been `#if / #endif / /*nothing*/`).  I pushed a hopefully correct version and I started a round of "precheck" CI.

**Comment by xavierleroy** [comment]
After rebasing and fixing a bit of this and that, "precheck" CI passed.  Let me merge this PR right now before it gets stale again.  In the near future I'll add a chi-square test on the results of `caml_sys_random_seed` to make sure that we're using `getentropy` or `/dev/urandom` correctly (unlike in my buggy commit 911d32aa6c85c73a3dfadd1fb5314d955ef1e779).

**Pull request merged**
